org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/ProjectOpenedOrClosedListener.java

```diff
@@ -81,102 +146,78 @@ public IBuilderState getBuilderState() {
 		return builderState;
 	}
 
-	@Inject
-	private IWorkspace workspace;
-
 	@Override
 	public void resourceChanged(final IResourceChangeEvent event) {
 		if (workspace != null && workspace.isAutoBuilding()) {
 			if (event.getType() == IResourceChangeEvent.POST_CHANGE) {
 				try {
-					final Set<IProject> toUpdate = Sets.newLinkedHashSet();
-					event.getDelta().accept(new IResourceDeltaVisitor() {
-						@Override
-						public boolean visit(IResourceDelta delta) throws CoreException {
-							if (delta.getResource() instanceof IWorkspaceRoot)
-								return true;
-							if (delta.getResource() instanceof IProject) {
-								IProject project = (IProject) delta.getResource();
-								if ((delta.getKind() & IResourceDelta.CHANGED) != 0 && project.isOpen()) {
-									if ((delta.getFlags() & IResourceDelta.OPEN) != 0) {
-										toUpdate.add(project);
-									}
-									if ((delta.getFlags() & IResourceDelta.DESCRIPTION) != 0) {	
-										if ((delta.findMember(new Path(".project")) != null)  
-												&& XtextProjectHelper.hasNature(project) && XtextProjectHelper.hasBuilder(project))
-											toUpdate.add(project);
-										else if(!XtextProjectHelper.hasNature(project)){
-											scheduleRemoveProjectJobIfNecessary(project, delta);
-										}
-									}
-								}
-							}
-							return false;
-						}
-					});
-					buildScheduler.scheduleBuildIfNecessary(toUpdate, IBuildFlag.FORGET_BUILD_STATE_ONLY);
+					final Set<IProject> toUpdate = findNewProjectsToBuild(event);
+					scheduleBuildIfNecessary(toUpdate);
 				} catch (CoreException e) {
 					log.error(e.getMessage(), e);
 				}
 			} else if ((event.getType() == IResourceChangeEvent.PRE_CLOSE || event.getType() == IResourceChangeEvent.PRE_DELETE)) {
-				if (event.getResource() instanceof IProject && 
-						(XtextProjectHelper.hasNature((IProject) event.getResource()) || 
-								event.getResource().isAccessible() && event.getResource().isHidden())) {
+				if (event.getResource() instanceof IProject && (XtextProjectHelper.hasNature((IProject) event.getResource())
+						|| event.getResource().isAccessible() && event.getResource().isHidden())) {
 					scheduleRemoveProjectJob((IProject) event.getResource());
 				}
 			}
 		}
 	}
 
-	protected void scheduleRemoveProjectJob(final IProject project) {
-		final ToBeBuilt toBeBuilt = getToBeBuiltComputer().removeProject(project, new NullProgressMonitor());
-		if (toBeBuilt.getToBeDeleted().isEmpty() && toBeBuilt.getToBeUpdated().isEmpty())
-			return;
-		new Job(Messages.ProjectOpenedOrClosedListener_RemovingProject + project.getName() + Messages.ProjectOpenedOrClosedListener_FromIndex) {
-			{
-				setRule(workspace.getRoot());
-			}
+	private Set<IProject> findNewProjectsToBuild(final IResourceChangeEvent event) throws CoreException {
```

Comment:
methods private beeing intentenional?

Reply:
One can override `visitResourceDelta` to customize the logic. I'm not passionate though and can make the others easily protected methods, too. What's your preference?